pkg/operator/controllers/machinehealthcheck/machinehealthcheck_controller_test.go

```diff
@@ -0,0 +1,157 @@
+package machinehealthcheck
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned"
+	arofake "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned/fake"
+	mock_dynamichelper "github.com/Azure/ARO-RP/pkg/util/mocks/dynamichelper"
+)
+
+// Test reconcile function
+func TestReconciler(t *testing.T) {
+	type test struct {
+		name    string
+		arocli  aroclient.Interface
+		mocks   func(mdh *mock_dynamichelper.MockInterface)
+		wantErr string
+	}
+
+	for _, tt := range []*test{
+		{
+			name: "Failure to get instance",
+			arocli: arofake.NewSimpleClientset(&arov1alpha1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "some other name",
+				},
+			}),
+			mocks:   func(mdh *mock_dynamichelper.MockInterface) {},
+			wantErr: `clusters.aro.openshift.io "cluster" not found`,
+		},
+		{
+			name: "Enabled Feature Flag is false",
+			arocli: arofake.NewSimpleClientset(&arov1alpha1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: arov1alpha1.SingletonClusterName,
+				},
+				Spec: arov1alpha1.ClusterSpec{
+					OperatorFlags: arov1alpha1.OperatorFlags{
+						enabled: strconv.FormatBool(false),
+					},
+				},
+			}),
+			mocks: func(mdh *mock_dynamichelper.MockInterface) {
+				mdh.EXPECT().EnsureDeleted(gomock.Any(), "MachineHealthCheck", "openshift-machine-api", "aro-machinehealthcheck").Times(0)
+				mdh.EXPECT().Ensure(gomock.Any(), gomock.Any()).Return(nil).Times(0)
+			},
+			wantErr: "",
+		},
+		{
+			name: "Managed Feature Flag is false: ensure mhc is deleted",
+			arocli: arofake.NewSimpleClientset(&arov1alpha1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: arov1alpha1.SingletonClusterName,
+				},
+				Spec: arov1alpha1.ClusterSpec{
+					OperatorFlags: arov1alpha1.OperatorFlags{
+						enabled: strconv.FormatBool(true),
+						managed: strconv.FormatBool(false),
+					},
+				},
+			}),
+			mocks: func(mdh *mock_dynamichelper.MockInterface) {
+				mdh.EXPECT().EnsureDeleted(gomock.Any(), "MachineHealthCheck", "openshift-machine-api", "aro-machinehealthcheck").Times(1)
+			},
+			wantErr: "",
+		},
+		{
+			name: "Managed Feature Flag is false: mhc fails to delete, an error is returned",
```

Comment:
It should retry to delete when the controller.Reconcile is called again. I think it would be worth looking into adding a limit on how many times the delete is retried and perhaps add a special log in the event we max out. I'll investigate this further in my local testing.

Reply:
I've added an hour wait between each requeue of the deletion attempt